### ISSUE TITLE
Open inside iframe in another site, pdf worker gets default url instantly

### ISSUE THREAD

**Issue by romanepifanov** [stephanrauh/ngx-extended-pdf-viewer#813]
```json
  "dependencies": {
    "@angular-material-components/datetime-picker": "^5.1.0",
    "@angular/animations": "~11.2.6",
    "@angular/cdk": "^11.2.5",
    "@angular/common": "~11.2.6",
    "@angular/compiler": "~11.2.6",
    "@angular/core": "~11.2.6",
    "@angular/forms": "~11.2.6",
    "@angular/material": "^11.2.6",
    "@angular/platform-browser": "~11.2.6",
    "@angular/platform-browser-dynamic": "~11.2.6",
    "@angular/router": "~11.2.6",
    "@ngrx/effects": "^12.1.0",
    "@ngrx/store": "^12.1.0",
    "classlist.js": "^1.1.20150312",
    "font-awesome": "^4.7.0",
    "ngx-cookie-service": "^11.0.2",
    "ngx-extended-pdf-viewer": "^9.0.2",
    "rxjs": "~6.6.0",
    "tslib": "^2.0.0",
    "web-animations-js": "^2.3.2",
    "zone.js": "~0.11.3"
  }
```
I have one application (angular + pdf viewer) there is pdf viewer working okay.
but if try to open this applications inside another by iframe
pdf.worker.js decides take default path. But other files loading right, how i set  in this option **pdfDefaultOptions.assetsFolder**
also i tried to use pdfDefaultOptions.workerSrc function, but it doesn't work.
  
![image](https://user-images.githubusercontent.com/25339462/123271484-8d7af680-d509-11eb-80d6-0c97ce9b8532.png)
![Screenshot 2021-06-24 163248](https://user-images.githubusercontent.com/25339462/123271790-d9c63680-d509-11eb-821a-c59c289ada9d.png)
![Screenshot 2021-06-24 163151](https://user-images.githubusercontent.com/25339462/123271792-da5ecd00-d509-11eb-89ab-d1a34310a8af.png)
![image](https://user-images.githubusercontent.com/25339462/123272435-67a22180-d50a-11eb-804a-23a30609e3af.png)

**Comment by stephanrauh** [comment]
First of all, I just saw this is your first ticket on GitHub ever. Welcome to GitHub in general and to the ngx-extended-pdf-viewer community in particular!

I'm not sure I've understood your issue correctly, so let me rephrase it in my own words:
- Your application is host in a context called - say - `app`.
- You've also mentioned that the application runs in an iFrame. I don't believe that makes a difference, but you never know. So there's a main application running in the root context, using an Angular application running in an iFrame. The iFrame content, in turn, is hosted in the `/app` context.
- As it turns out, almost every file of ngx-extended-pdf-viewer is loaded correctly from the `/app` context - with the sole exception of pdf.worker-*.js, which is loaded from the root context.
- Modifying the attribute `pdfDefaultOptions.workerSrc` didn't make a difference. The worker file is still loaded from the root context.

**Comment by stephanrauh** [comment]
If you're in a hurry (or desperate), you can load the worker file manually. I don't recommend that, because the file name changes with almost every release and because the performance of the PDF viewer suffers considerably. 

But if you don't mind that, simply add the file `pdf-2.9.*.min.js` to the `scripts` section of your `angular.json`. This means the worker file is loaded when the application starts, so ngx-extended-pdf-viewer won't try to load it from the wrong folder.

**Comment by stephanrauh** [comment]
If you've got a minute to spare and if you want to help me, try debugging the library. I'd appreciate that because I'm not sure I'm able to write a reproducer. Here's what to do:

- Add a `debugger` or `console.log` statement to your custom implementation of `pdfDefaultOptions.workerSrc`. I wonder if your implementation is called at all?
- If it's not called, maybe it's a simple timing problem. You have to set `pdfDefaultOptions.workerSrc` *before* ngx-extended-pdf-viewer is loaded. That's why I recommend setting the `pdfDefaultOptions` in the module file.
- If you're still struck, set `useMinifiedJSLibraries="false"` and add a breakpoint here: https://github.com/stephanrauh/pdf.js/blob/d083a05720434de29d59de6e413043c022ee1c1a/src/display/api.js#L2079
- That's the line responsible for loading the worker file. I suppose it's got the wrong path for some reason. If you can confirm that, tell me. Maybe you can even find out what's going on. That'd be great because it'd allow me to fix the bug in a few minutes (if it's a bug). 

I know I'm demanding a lot - but even a partial analysis would help me to help you!

Thanks in advance,
Stephan

**Comment by romanepifanov** [comment]
I have one application with pdf viewer
and another proxy application with iframe(where i show the first application)
If i open a pdf in first applications, all works great
if i open it by proxy, somethin is changing path of the pdf worker

the first screenshotof executing inside first application
![image](https://user-images.githubusercontent.com/25339462/123430121-7d7a1a00-d5d0-11eb-9a1e-f1f9172be17a.png)
the second screenshotof executing inside iframe
![image](https://user-images.githubusercontent.com/25339462/123430307-b4503000-d5d0-11eb-88a3-37f744010d63.png)

but after call the `pdfDefaultOptions.workerSrc` it anyway returns wrong path
u can see it inside 3rd screenshot
![image](https://user-images.githubusercontent.com/25339462/123430570-042ef700-d5d1-11eb-9133-2813059431db.png)

**Comment by romanepifanov** [comment]
I have one application with pdf viewer
and another proxy application with iframe(where i show the first application)
If i open a pdf in first applications, all works great
if i open it by proxy, somethin is changing path of the pdf worker

the first screenshotof executing inside first application
![image](https://user-images.githubusercontent.com/25339462/123430121-7d7a1a00-d5d0-11eb-9a1e-f1f9172be17a.png)
the second screenshotof executing inside iframe
![image](https://user-images.githubusercontent.com/25339462/123430307-b4503000-d5d0-11eb-88a3-37f744010d63.png)

but after call the `pdfDefaultOptions.workerSrc` it anyway returns wrong path
u can see it inside 3rd screenshot
![Screenshot 2021-06-25 162030](https://user-images.githubusercontent.com/25339462/123430807-46f0cf00-d5d1-11eb-85b9-4487345ac917.png)

**Comment by romanepifanov** [comment]
Inside viewer i see the hardcoded path
![image](https://user-images.githubusercontent.com/25339462/123431628-2412ea80-d5d2-11eb-8614-d46c74e93b85.png)
these is expression for checking use default or custom, but how i understood it is always string, thats why it returns the default path of the pdf worker
![image](https://user-images.githubusercontent.com/25339462/123432247-d8ad0c00-d5d2-11eb-882d-c2db57492852.png)
![image](https://user-images.githubusercontent.com/25339462/123432790-67ba2400-d5d3-11eb-9d8a-5c7a03b9b2d1.png)

**Comment by stephanrauh** [comment]
Weird. Can you send me a reproducer? A small but complete Angular application showing the bug?

I suspect the Angular application in the iFrame is initialized incorrectly. If I get your last screenshot right, `_worker.options.GlobalWorkerOptions.workerSrc` is a string. But that's only possible if ngx-extended-pdf-viewer does not initialize correctly.

**Comment by stephanrauh** [comment]
This `for` loop is reponsible for setting the `workerSrc` function. I can only guess, but maybe it's never called in your iFrame.

![image](https://user-images.githubusercontent.com/3045767/123470962-66015800-d5f5-11eb-8168-bf6a1532b691.png)

**Comment by stephanrauh** [comment]
I've found something that looks suspiciously like a bug. Before fixing the bug I'd like to check if the bug fix breaks existing applications.

https://github.com/stephanrauh/pdf.js/blob/d083a05720434de29d59de6e413043c022ee1c1a/src/display/api.js#L243 and https://github.com/stephanrauh/pdf.js/blob/d083a05720434de29d59de6e413043c022ee1c1a/src/display/api.js#L2073 look wrong. This version of `new URL()´ ignores the last part of the base URL. That'd explain your bug.

**Comment by romanepifanov** [comment]
Hello, sorry for the delay. We decide to use default pdf.js. If you want you can close this issue

**Comment by stephanrauh** [comment]
Yeah, but pdf.js suffers from the same bug, doesn't it?

**Comment by romanepifanov** [comment]
Maybe yes, but i overwrite script, whitch takes the worker file. We decide to do it, becasue we need support for IE 11 and only pdf.js v1.10.100 can do it

**Comment by stephanrauh** [comment]
I see. IE11 is always a good reason. Good luck!

I'll fix the bug I've found nonetheless. I'm just waiting until I've collected enough stuff to publish a new major version (10.0.0).

**Comment by stephanrauh** [comment]
Reproducer: https://github.com/stephanrauh/ngx-extended-pdf-viewer-issues/tree/main/issue813-path-plus-iframe

Start it with "npm start". The reproducer serves the Angular application from the path localhost:4200/angularapp/. The PDF file uses a relative path (`./assets/my.pdf`).

Observed behavior:

- `http://localhost:4200/angularapp` doesn't find the PDF file.
- `http://localhost:4200/angularapp/` works. (Note the trailing slash).
- `http://localhost:4200/angularapp/page1` works.
- `http://localhost:4200/angularapp/chapter1/page1` does not work.

**Comment by stephanrauh** [comment]
Problem analysis: `new URL(workerSrc, window.location).href` can't distinguish between folders and resource files. Plus, it doesn't play well with the router URLs of Angular. `http://localhost:4200/angularapp/chapter1/page1` is a virtual path. It's better to use `BASE_HREF` instead of `window.location`.

**Comment by stephanrauh** [comment]
The bugfix has shipped with version 10.0.0-alpha.11.

**Comment by kmr24** [comment]
in this  10.0.0-alpha.11 version if do a production build using npm build --configuration production then entire supporting files are getting generated in asset directory . see the attached picture

![image](https://user-images.githubusercontent.com/90660267/133173953-2867a9f3-1394-4a65-b03d-8a1b867f2c9b.png)

**Comment by stephanrauh** [comment]
I suppose you want to get rid of the *.map files? And maybe even of the non-minified files?

I'm afraid there's nothing I can do about it. But you can. It's just a bit cumbersome. You've configured the `angular.json` to copy every file from `node_modules/ngx-extended-pdf-viewer/assets` to your `assets` folder. Modify the glob rules to exclude the files you don't need. Unfortunately, excluding doesn't work for some reason, so you need to specify every file pattern and every folder you want to keep.
